File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        def traverse(i , j):
            if obstacleGrid[i][j] == 1 : return 0
            if i ==0 and j == 0: return 1
            if i == -1 or j == - 1: return 0
            if dp[i][j] != -1 : return dp[i][j]
            left = traverse(i , j-1)
            up = traverse(i-1 , j)
            dp[i][j] = left + up
            return left + up
        dp = [[-1 for i in range(len(obstacleGrid[0]))] for j in range(len(obstacleGrid))]
        return traverse(len(obstacleGrid)-1 , len(obstacleGrid[0])-1)

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int traverse(int i , int j , vector<vector<int>>& obstacleGrid , vector<vector<long long>> &dp){
        if (i<0 || j<0) return 0;
        if (obstacleGrid[i][j] == 1)  return 0;
        if (i==0 && j == 0) return 1;


        if (dp[i][j] != -1) return dp[i][j];
        int up =0 ;
        int left = 0;
        up = traverse(i-1 , j , obstacleGrid , dp);
        left = traverse(i , j-1 , obstacleGrid , dp);
        dp[i][j] = up+left;
        return up + left ;
    }
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        vector<vector<long long>> dp(obstacleGrid.size() , vector<long long>(obstacleGrid[0].size() , -1)) ;
        return traverse(obstacleGrid.size()-1 , obstacleGrid[0].size()-1 , obstacleGrid , dp);
    }
};